Non rotated and rotated backgrounds

illudedPerception opened the pull request:
* Added height numbers to Mesa Plains, height numer accuracy is ~2-3m cause of rounding and raster data manipulation
* Added back non rotated images (reverted to live version images)
* Rotated images have "180" in the end

Transform data for rotated images and these were tested with this https://github.com/embix/railroad.studio/tree/spike-merge-illu-pr1 unless theres big changes, this transform data is valid and checked against my own save games

MesaPlainsTopo.png
140.89,0,0,140.89,-300899.87,-300459.45

AuroraFalls180.png
139.91,0,0,139.91,-300460.04,-300900.04

LakeValley180.png
91.88,0,0,91.88,-197454.05,-197454.05

PineValley180.png
92.95,0,0,92.95,-199953.52,-199953.52

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Updated background rendering for different game levels, introducing new visual representations for 'MesaPlains', 'AuroraFalls', and other game maps.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

illudedPerception committed:
Update RailroadMap.ts

illudedPerception committed:
images

scottanderson commented on ts/RailroadMap.ts:
```diff
@@ -744,17 +744,18 @@ export class RailroadMap {
-            transform = 'matrix(-93.16,0,0,-93.16,197453,197453)';
-            // image = 'LakeValleyMap2.png';
-            // transform = 'matrix(-200,0,0,-200,200000,200000)';
+        if (this.railroad.settings.gameLevelName === 'MesaPlains') {
```
Please do not reorder the maps, this makes the change harder to review. You can add MesaPlains before else.

illudedPerception replied:
my bad on this one

scottanderson reviewed (changes requested):
Thanks for the contribution, @illudedPerception! A couple of points to address:

1. There's no need to open a new PR for each edit—you can simply update the existing branch. This will help to keep all related feedback in a single PR, making it easier to reference comments on older revisions.

2. It looks like feedback from the previous PR has not been addressed. This PR still renames existing files and touches TS code, which will create problems with CDN caching. Let's not modify the file names, and let's not make any changes to the ts code in this change, since the new maps cannot be used until the other PR is ready.

Let me know if you need further clarification!

illudedPerception committed:
Update RailroadMap.ts

illudedPerception commented:
> 
>     2. It looks like feedback from the previous PR has not been addressed. This PR still renames existing files and touches TS code, which will create problems with CDN caching. Let's not modify the file names, and let's not make any changes to the ts code in this change, since the new maps cannot be used until the other PR is ready.

*reverted back to the live TS file
no files what are in the live version have been renamed in this pr, just added 3 rotated images (as requested) just with different names and Mesa plains does not have a rotated image because this map was added post rotation of the in game maps
![image](https://github.com/user-attachments/assets/fa0eacdb-efb1-497a-9e3a-252c35bccaf6)

illudedPerception commented:
![image](https://github.com/user-attachments/assets/de81fa7c-6cd2-4206-980e-3c466e3c52b7)
![image](https://github.com/user-attachments/assets/8ffa61d4-e14e-4e39-a0d4-9c6dfb92d5e4)

illudedPerception reviewed (commented):
reverted to the live TS
no files were renamed in this pr, just added 3 new rotated images with "180" instead of "topo" in the end of the names

scottanderson reviewed (changes requested):
This PR still modifies the original files. This will break all of the modified maps, because the transform doesn't match the modified image.

The new files are also not the correct size.

illudedPerception committed:
revert to live images

illudedPerception commented:
> This PR still modifies the original files. This will break all of the modified maps, because the transform doesn't match the modified image.
* Reverted back to live version images

> 
> The new files are also not the correct size.
* Image size will change a bit after editing the map files in anyway, that's why i all ways update the transforms, what are exported from QGIS with the images and like i have said before exporting from QGIS works on zoom level of the view port, as i mentioned and demostrated here https://github.com/scottanderson/railroad.studio/pull/246#issuecomment-2562429446

scottanderson commented:
Thank you for the effort you've put into this PR. While I see that you're addressing the incremental feedback provided, the broader requirements I’ve outlined previously remain unresolved. Unfortunately, we’re not converging on a solution that meets the project’s standards for asset handling, The current approach with raster images embeds too many assumptions that cannot be adjusted without regenerating the images—something only you are currently able to do. This significantly limits RS’s ability to work with topographical data, it limits the resolution of the background images at high zoom, and it consumes more bandwidth than is necessary because everyone has to download the full-resolution raster image. For these reasons, I'm going to decline this PR.

I see two viable paths forward:

1. Transition to vector graphics for the topographical layer.
2. Provide a documented, reproducible process for creating the raster images, ensuring that anyone can generate identical assets consistently.

Until we adopt vector-based assets or establish a reproducible workflow for raster images, this approach cannot proceed.

I hope this clarifies the situation, and I’m happy to collaborate on different approaches if needed.

illudedPerception commented:
>     1. Transition to vector graphics for the topographical layer.
>
>     2. Provide a documented, reproducible process for creating the raster images, ensuring that anyone can generate identical assets consistently.

1. There is nothing consistent in any of the methods of getting information from the game at this point, because like i said RROx was discontinued so that option is gone and the other ways also rely on community made tools what might be discontinued at some point as well and there are other reasons why not to use the other tools, even RROx was against the TOS of Ro and the other tools ain't so compliant either

2. There is an option to cut out most of the tools and make the mod loader (RE-UE4SS) use a line trace script (like RROx did) so porting the RROx line trace code to LUA script and use it with the mod loader would be a start and the mod loader has a line trace mod for name tracing of in game actors and its very similar to RROx line trace code for cords. Porting the code is above my head, i just understand enough to see that its similar. Still this method relies on a tool made by other people what might close at some point or might not.

after overcoming the problem of getting the data from the game, we still get to the SVG and consistent size problem and that's the current problem i don't have a solution or an idea how to even start to solve it.

It was great to give to the RO community as i could in the extent of my skills and i accuired some new ones on the way, but i'm sad to say that seems this is the end for me on this project, i wish you good luck and lets hope some one smarter comes to work on the images from here forward
@scottanderson i thank you for the opportunity of letting me contribute to this project.

scottanderson commented:
I'm sorry that we do not see eye to eye on this, but I feel that this dialogue has not been constructive.

1. The inverted map cannot be a different size that the non inverted map.

2. I asked you to share the topographical data files - not to describe how you created them.

3. I asked you to document the process for converting topographical data to raster images so that others besides yourself may contribute to this area.

If we had #3, there would be a reason for someone to collect topo data for Mesa Plains, but without having a way to convert topi data to an image, there is no reason for anyone to collect the data.

I hope this helps you to see the bigger picture here and why I am pushing for a repeatable process, but If you still want to take your ball and go home, that's your choice. I have made every effort to explain why I asked for changes, and to provide you with clear instructions on how you can fix the issues, but I am not seeing a reciprocal effort from you to address my feedback, and I don't want to waste any more of your time on making edits that that are not aligned with the project's goals.

illudedPerception commented:
>     1. The inverted map cannot be a different size that the non inverted map.

There are other ways to export images in the way of "print layout". 

https://github.com/illudedPerception/RO.topoData/tree/main/test has two test exports one SVG and the other PNG both exported via print layout (not the current way of exporting) and no editing done to them though the border on both of them is big and that might be a problem, but that's the downside of that style of exporting, upside is consistent picture sizes (higher DPI might result in better quality on the png)

>     2. I asked you to share the topographical data files - not to describe how you created them.

Well here they are, though they are outdated because i generated them before the rotated maps, so new ones would be in order
https://github.com/illudedPerception/RO.topoData

>     3. I asked you to document the process for converting topographical data to raster images so that others besides yourself may contribute to this area.

Read me is a long process and writing it is error prone, allready because of the amount of screenshots id need and me writing it all, i might work on a quick youtube video showing how i do it now with the CSV files if thats okay

illudedPerception commented:
* The data is here, generated pre rotation so its  kinda outdated https://github.com/illudedPerception/RO.topoData
* There is a test folder and it includes a PNG exported by "print layout" not the way i export images, mostly cause of the borders it leaves and the problems it would bring for lining up the save game data with the raster, borders what can't be removed pre exporting, at least not in a way it would leave repeatable image size or other aspects the same with each image export
*My bet is that the SVG file exported from QGIS is not what you would want to use or its just me thinking that, either way you can take a look at it in the test folder, if this svg is not what you want then QGIS is not the program to continue on the vector image path
* I could show how to generate a raster in QGIS but why add more videos into the mix, there's many videos showing how to do it. Also i'm not gonna write half a bible long readme file with screen shots and then there being some errors what result in no raster generated or someone editing the parameters so it would generate faster or something in those lines, also to get the images what are live now would need hours of manual clean up in layer editing inside QGIS (to get a cleaner not so crowded image, also the editing needs to be thought thru) and all of that only if a CSV file was used, raster based data needs alot of work before it even is ready to be generated into a base raster what could be used to generate the contour lines what has too many steps and those steps can have changes from person to person and even if i make 10 project files with the same raster the end result might be different every time. 
I might mess around with the "print layout" exporting method to get to a point where i can export images with a standard, what works everytime

scottanderson commented:
Thank you for sharing this information, it’s a valuable step forward.

My goal is not to teach others how to manually create the raster images, but to automate their generation. This will allow collaborators to modify the process itself by editing the scripts that generate the files, without needing to replicate the extensive manual effort you’ve described.

This will make it easier to review—and therefore to accept—changes to these assets.

To move toward automation, we just need a clear description of the steps involved in generating these images. I believe this is what you’ve provided with the linked repository, although it will take me some time to fully understand its contents.

I haven’t used QGIS extensively myself, but I understand that the desktop application acts as a wrapper around the QGIS-Processing API, which is also accessible via the CLI. After configuring the settings in the desktop app initially, you should be able to copy the QGIS-Processing `run` commands for use in a script. Similarly, converting SVG to raster can be automated with available tools.

I don’t expect you to write the script yourself, but if you could provide the Processing run commands you used to convert CSV to SVG, I could put something together for that.

I want to emphasize how much I appreciate the time and effort you’ve put into this project. Your contributions have been instrumental in advancing it, and I hope we can continue collaborating to find a solution that benefits the community.

The decision to close this PR is not a reflection of the value of your contributions, but rather a step toward aligning the project with its long-term goals. Your skills and perspective have been incredibly valuable, and I hope you’ll consider continuing to contribute in the future. There’s still a lot we can accomplish together, and your expertise would be a tremendous asset as we work to make this process more accessible and repeatable for others.

illudedPerception commented:
*The manual work is to enable artifact free image but let that be a future problem
1. I open the CSV with data source manager 
![image](https://github.com/user-attachments/assets/8c4bd9e4-b0ad-4e22-b268-29147c9282e6)
2. Then i use Tin interpolation from the processing tool bar
* interpolation attribute Z
* press the green cross and set to points (as in the image)
* pixel size 100 rows and columns set automatically
* set the extent by calculate from layer
* also i save the interpolated raster to file (maybe not needed for your process) but its needed for mine
![image](https://github.com/user-attachments/assets/e29444ed-f912-4c77-b0cd-11b31152a88d)
![image](https://github.com/user-attachments/assets/d9a5a23d-1ca1-47e9-a404-74393f7604b2)
3. Then i hover to open the raster menu on the top and select extraction then contour
![image](https://github.com/user-attachments/assets/40df275f-2116-4075-9811-021e5898779a)
* Set the interval to 500 (aka 5m)
* Offset 479 (for aurora falls) its where the contours start at or at least how i understand it and 479 is also the lowest part in the map the player can walk on or the line trace can hit
![image](https://github.com/user-attachments/assets/660f8465-a235-489d-8c38-57b2fec1b38a)
4. Then the field calculator i really can't help here cause its been a bit different with each map, but its still needed to make the different calculations like dividing the ELEV to have a out put in meters not in cm and to assign different height values to two groups if i remember right i had one index column and two label columns
something something this here its from my own "readme file", you as a coder prob can streamline the butchery written below :D
Field calculator (might need editing)
To make index:					if("ELEV"//100%10=0, 1, NULL)
To make index 2:				if("ELEV"//100%10=0, NULL, 2)
To make label:					round("index"*"elev"/100,0)
To make label 2:				round("index2"/2*"elev"/100,0)

* This is the problem i had/still kinda have with Mesa plains the band was a scale of gray, not a scale of Z height transfered from the CSV as in this picture below
![image](https://github.com/user-attachments/assets/b4ff95dd-d56a-47d4-b814-28ebe392ea02)

illudedPerception commented:
* The artifacts i talk about
![image](https://github.com/user-attachments/assets/00a01c1f-51df-48c4-b839-d625287a187c)
![image](https://github.com/user-attachments/assets/4ceeea5a-7e58-499c-8851-73eb13c79e1a)
* step by step i zoom in so you would get the idea why i edited out alot of it
![image](https://github.com/user-attachments/assets/265ed9a0-1f06-4eb7-bc67-d99e42fa237b)
![image](https://github.com/user-attachments/assets/4bbd2cb7-3c36-42e8-8b16-61b718174d5e)
![image](https://github.com/user-attachments/assets/f06aba8e-4c2d-41e6-90f9-ff4d63df00d8)

illudedPerception commented:
Sorry i rushed a bit il add the commands here or the ones i can
* tin interpolation
`processing.run("qgis:tininterpolation", {'INTERPOLATION_DATA':'file:///C:/Users/raini/Desktop/PineValley.csv?type=csv&maxFields=10000&detectTypes=yes&xField=X&yField=Y&zField=Z&crs=EPSG:4326&spatialIndex=yes&subsetIndex=no&watchFile=no::~::1::~::-1::~::0','METHOD':0,'EXTENT':'-200000.000000000,200000.000000000,-200000.000000000,200000.000000000 [EPSG:4326]','PIXEL_SIZE':100,'OUTPUT':'TEMPORARY_OUTPUT'})`
* contours
`processing.run("gdal:contour", {'INPUT':'C:/Users/raini/AppData/Local/Temp/processing_TDmxzM/369142e141aa4e5fb8e1a34f911ffae3/OUTPUT.tif','BAND':1,'INTERVAL':500,'FIELD_NAME':'ELEV','CREATE_3D':False,'IGNORE_NODATA':False,'NODATA':None,'OFFSET':4039,'EXTRA':'','OUTPUT':'C:/Users/raini/Desktop/contours.shp'})`

for exporting the image file, there is not a place to copy the command

scottanderson commented:
I get an error when I try to run the tininterpolation step. I tested on Windows and Ubuntu, but I got the same error both times. What version of QGIS are you using? Do I need to install any other dependencies?

```
+ $ qgis_process run qgis:tininterpolation '--INTERPOLATION_DATA=topo CSV/PineValley.csv?type=csv&maxFields=10000&detectTypes=yes&xField=X&yField=Y&zField=Z&crs=EPSG:4326&spatialIndex=yes&subsetIndex=no&watchFile=no::~::1::~::-1::~::0' --METHOD=0 '--EXTENT=-200000.000000000,200000.000000000,-200000.000000000,200000.000000000 [EPSG:4326]' --PIXEL_SIZE=100 --OUTPUT=output.tif

----------------
Inputs
----------------

EXTENT: -200000.000000000,200000.000000000,-200000.000000000,200000.000000000 [EPSG:4326]
INTERPOLATION_DATA:     topo CSV/PineValley.csv?type=csv&maxFields=10000&detectTypes=yes&xField=X&yField=Y&zField=Z&crs=EPSG:4326&spatialIndex=yes&subsetIndex=no&watchFile=no::~::1::~::-1::~::0
METHOD: 0
OUTPUT: output.tif
PIXEL_SIZE:     100


Logged critical: Traceback (most recent call last):
  File "/usr/share/qgis/python/plugins/processing/algs/qgis/TinInterpolation.py", line 153, in processAlgorithm
    crs = layer.sourceCrs()
          ^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'sourceCrs'

Traceback (most recent call last):
  File "/usr/share/qgis/python/plugins/processing/algs/qgis/TinInterpolation.py", line 153, in processAlgorithm
    crs = layer.sourceCrs()
          ^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'sourceCrs'

ERROR:  Traceback (most recent call last):
  File "/usr/share/qgis/python/plugins/processing/algs/qgis/TinInterpolation.py", line 153, in processAlgorithm
    crs = layer.sourceCrs()
          ^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'sourceCrs'

Traceback (most recent call last):
  File "/usr/share/qgis/python/plugins/processing/algs/qgis/TinInterpolation.py", line 153, in processAlgorithm
    crs = layer.sourceCrs()
          ^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'sourceCrs'
```

illudedPerception commented:
> I get an error when I try to run the tininterpolation step. I tested on Windows and Ubuntu, but I got the same error both times. What version of QGIS are you using? Do I need to install any other dependencies?


this was all done on the vanilla processing tools in QGIS and im on 3.40.2 'Bratislava' and then i used the copy as pyhton command i usually don't run them with the commands so i don't know if i got the right one

illudedPerception commented:
interpolation but copied as wgis prosess run instead of python

`qgis_process run qgis:tininterpolation --distance_units=cm --area_units=cm2 --ellipsoid=EPSG:7030 --INTERPOLATION_DATA='file:///C:/Users/raini/Desktop/AuroraFallsXYZ.csv?type=csv&maxFields=10000&detectTypes=yes&xField=X&yField=Y&zField=Z&crs=EPSG:4326&spatialIndex=yes&subsetIndex=no&watchFile=no::~::1::~::-1::~::0' --METHOD=0 --EXTENT='-300970.000000000,300530.000000000,-300970.000000000,300530.000000000 [EPSG:4326]' --PIXEL_SIZE=100 --OUTPUT=TEMPORARY_OUTPUT`

and extracting one

`qgis_process run gdal:contour --distance_units=cm --area_units=cm2 --ellipsoid=EPSG:7030 --INPUT='C:/Users/raini/AppData/Local/Temp/processing_ecFOPs/f0ad1d49369749b380ce5a425423f802/OUTPUT.tif' --BAND=1 --INTERVAL=500 --FIELD_NAME=ELEV --CREATE_3D=false --IGNORE_NODATA=false --OFFSET=-479 --EXTRA= --OUTPUT='C:/Users/raini/Desktop/contour.shp'`

scottanderson commented:
I upgraded from LTS to Bratislava, but I am still getting the same error. I also tried the docker version of qgis, but got the same results with that as well.

The latest release of https://github.com/ccrook/QGIS-Contour-Plugin claims to fix the error I'm seeing, but it's not clear how to install qgis plugins.

scottanderson commented:
1. I was able to generate the heightmap tiff in the desktop app, but have not had any luck with doing it from the cli. I suspect that it only works through python, but it will take me some time to test that.

2. I was able to convert tif to shp using the cli. It even works with the docker container.
```
$ docker run -it --rm -v $(pwd):/data -e QT_QPA_PLATFORM=offscreen qgis/qgis:latest gdal_contour -b 1 -a ELEV -i 500.0 -off -1.0 -f "ESRI Shapefile" /data/interpolated.tif /data/contour.shp
0...10...20...30...40...50...60...70...80...90...100 - done.
```

3. I'm not sure how to convert the contour shp to the final png. Can you elaborate on how this is done in the desktop app?

illudedPerception commented:
>     3. I'm not sure how to convert the contour shp to the final png. Can you elaborate on how this is done in the desktop app?
Theres a two ways to do it
1. i use the import/export feature under project menu
* but the downside of this is the inconsistent PNG size
![image](https://github.com/user-attachments/assets/a8a7e987-ba0a-44a6-bb54-f5ff62bde951)
![image](https://github.com/user-attachments/assets/a14935fe-2ff4-4888-9343-269238db92ac)
2. is the print layout there's some setup though
* The downside is it leaves a big border what means the image needs to be bigger then it is now to retain the detail we have now and the world file cant be used, unless we use the work around i thought of yesterday
 SO..
* First i open the project menu and select "New print layout" or ctrl+p
* Then a small window opens and asks for a name, name of the layout seems not to be important
* A new window opens, and i right click on the page and settings options will show at bottom right (at least for me)
![image](https://github.com/user-attachments/assets/843aafa5-c790-48db-90d8-87ca4caa2a21)
![image](https://github.com/user-attachments/assets/440d05b9-ef89-4523-8074-ca5e90b73b93)
* Then i set the page size to 300x300 anything over that seems to impact the readability of the height numbers
* and set the background color to #ffffec by clicking the dropdown menu arrow of the color selector
* then i select add map from the left side menu and let the drawing tool snap to the edges of the page and draw a square on the page can be resized later
![image](https://github.com/user-attachments/assets/ce16f126-2839-4c52-abe9-0c0e834d265e)
end result is this
![image](https://github.com/user-attachments/assets/e4b4e97e-b337-404f-8abe-c0a3caee7718)
* Then i use export image, it asks a location to save the file to
![image](https://github.com/user-attachments/assets/6d76122f-5cc6-48ba-956e-c41e204a5214)
* Then change the size as needed DPI will adjust automatically
* The world file is the transform data, but because of the borders its not gonna help to match the save game data to the image
(i have an idea for fixing, that problem but its again manual work)
![image](https://github.com/user-attachments/assets/bb72a0d0-4685-41b2-b336-84c3abb81dcc)

illudedPerception commented:
also the image size is affected by the map extent so maps with similar sizes come out the same image size, pine valley is the smallest map, because of how the world borders are placed

illudedPerception commented:
well print layout is also getting the map to the page by the view port size, so if it has borders on the side it adds the same size borders all around
![image](https://github.com/user-attachments/assets/da75e20b-bcea-40d2-8c97-c11946626e97)
* This is my current viewport, now as i drag the right side processing toolbar to the right the border in the print layout will get bigger
![image](https://github.com/user-attachments/assets/115f9fa9-d39a-469c-8300-6a6fc8600a2e)
* now this made me think that about the placement of the map in the view port and if it would affect the world file if i export the image through print layout

illudedPerception commented:
as i found out print layout is also in the processing toolbar as a script what could be used, but it can't be used without a print layout what can't be made with a script at least i didn't find one for making a print layout and any print layout template made before is specific to each map because the extent of the map is saved to the template, so it would not work for new/other maps, cause each map has their borders placed a bit different making the extent different as well, plus the coordinates extraction extent was set by hand into RROx's heightmapper.ts and then re compiled and loaded as dev plugin

illudedPerception commented:
yes and it works as i thought
![image](https://github.com/user-attachments/assets/af5f676a-0235-4f64-b434-d803b5a378a8)
![image](https://github.com/user-attachments/assets/dfd32a84-cfad-4846-9a2f-92480a039850)
* and it affects the world file too, still gonna run one more test to see how consistent i can get the image (not meant in size) but as do the features stay in the same place if the export settings are same but the map is moved in the QGIS main window
* Result on the test is that the position in QGIS main window affects the features by 1px (not in image size) what in my mind is nothing, also the zoom in the main window affects the size of the map in print layout what i guess also changes the world file aka the transform, seems automation is the way unless the kinks cant be worked out
![image](https://github.com/user-attachments/assets/48667e15-8d7c-48cb-b6f2-794ac66f71f4)
